### PR TITLE
Add a nice wrapper for `get_sys_info`

### PR DESCRIPTION
**Pull request by nahla-nee**
Implements a nice safe wrapper for `get_sys_info` as described in #839 .

### CODE REVIEW

**Review by thejpster — CHANGES_REQUESTED** [review]
I'd like to see an example of using this API.

**Comment by nahla-nee** [comment]
Sure! And I'll go ahead make the suggestions changes too. I'll probably do that by tonight (us time). Do you want the example to be added to the existing rom function example? And if so do you want it to replace the existing one?

**Comment by thejpster** [comment]
I'm ok with it going into the existing example binary. I'm not sure if we need to leave an example of the old API or if we expect people to only use the new one.

File: rp235x-hal/src/rom_data.rs [hunk 1]
```diff
@@ -109,6 +109,435 @@ pub fn rom_data_lookup(tag: RomFnTableCode, mask: u32) -> usize {
...
+            -4 => Self::NotPermitted,
+            -5 => Self::InvalidArg,
+            -19..=-10 => unsafe { core::mem::transmute::<i32, BootRomApiErrorCode>(value) },
+            _ => Self::Unknown,
+        }
+    }
+}
+
+#[allow(unused)]
+mod sys_info_api {
```
[the remark this answers is not in the document]
Comment (nahla-nee) [reply rp235x-hal/src/rom_data.rs]: My bad, can't believe I missed that 😅 Fixed.

**Comment by nahla-nee** [comment]
I also went ahead and added some more documentation. It probably could use more work cause I was in a bit of a rush but it should make things easier to read. I also elaborated on the existing examples since we don't have settle for printing the hex representation of the bytes get_sys_info gives back now.

File: rp235x-hal/src/rom_data.rs [hunk 2]
```diff
@@ -109,6 +109,526 @@ pub fn rom_data_lookup(tag: RomFnTableCode, mask: u32) -> usize {
...
+        /// The flag used to get the random 128 bit integer generated on boot
+        BootRandom = 0x0010,
+        // Ignore nonce for now since it can't/shouldn't be called anyway?
+        // Nonce = 0x0020,
+        /// The flag used to get boot diagnostic info
+        BootInfo = 0x0040,
+    }
+
+    impl GetSysInfoFlag {
+        const fn buffer_length(&self) -> usize {
```
Comment (thejpster) [inline rp235x-hal/src/rom_data.rs:197]: Maybe add a note that the sys info flags are a bitmask and can be OR'd together, but we only ever send one flag at a time so we don't need to add the sizes together.

File: rp235x-hal-examples/src/bin/rom_funcs.rs [hunk 3]
```diff
@@ -234,31 +235,58 @@ fn get_sys_info_chip_info<T>(uart: &mut T)
...
     T: core::fmt::Write,
 {
-    let mut buffer = [0u32; 16];
-    let result = unsafe { hal::rom_data::get_sys_info(buffer.as_mut_ptr(), buffer.len(), 0x0001) };
-    _ = writeln!(uart, "get_sys_info(CHIP_INFO/0x0001) -> {}", result);
-    _ = writeln!(uart, "\tSupported Flags: {:#06x}", buffer[0]);
-    _ = writeln!(uart, "\tRP2350 Package ID: {:#010x}", buffer[1]);
-    _ = writeln!(uart, "\tRP2350 Device ID : {:#010x}", buffer[2]);
-    _ = writeln!(uart, "\tRP2350 Wafer ID  : {:#010x}", buffer[3]);
+    let result = hal::rom_data::sys_info_api::chip_info();
```
Comment (thejpster) [inline rp235x-hal-examples/src/bin/rom_funcs.rs:238]: These nested matches can be collapsed.

```rust
let chip_info = match hal::rom_data::sys_info_api::chip_info() {
    Ok(Some(info)) => info,
    Ok(None) => {
        _ = writeln!(uart, "chip_info() not supported");
        return;
    }
    Err(e) => {
        _ = writeln!(uart, "Failed to get chip info : {:?}", e);
        return;
    }
};
```

Are we expecting chips to not support chip info?

File: rp235x-hal-examples/src/bin/rom_funcs.rs [hunk 4]
```diff
@@ -234,31 +235,58 @@ fn get_sys_info_chip_info<T>(uart: &mut T)
...
 fn get_sys_info_cpu_info<T>(uart: &mut T)
 where
     T: core::fmt::Write,
 {
-    let mut buffer = [0u32; 16];
-    let result = unsafe { hal::rom_data::get_sys_info(buffer.as_mut_ptr(), buffer.len(), 0x0004) };
-    _ = writeln!(uart, "get_sys_info(CPU_INFO/0x0004) -> {}", result);
-    _ = writeln!(uart, "\tSupported Flags: {:#06x}", buffer[0]);
+    let result = hal::rom_data::sys_info_api::cpu_info();
+    let result = match result {
```
Comment (thejpster) [inline rp235x-hal-examples/src/bin/rom_funcs.rs:265]: These nested matches can be collapsed.

File: rp235x-hal-examples/src/bin/rom_funcs.rs [hunk 5]
```diff
@@ -268,56 +296,116 @@ fn get_sys_info_flash_dev_info<T>(uart: &mut T)
 where
     T: core::fmt::Write,
 {
-    let mut buffer = [0u32; 16];
-    let result = unsafe { hal::rom_data::get_sys_info(buffer.as_mut_ptr(), buffer.len(), 0x0008) };
-    _ = writeln!(uart, "get_sys_info(FLASH_DEV_INFO/0x0008) -> {}", result);
+    let result = hal::rom_data::sys_info_api::flash_dev_info();
+    let result = match result {
```
Comment (thejpster) [inline rp235x-hal-examples/src/bin/rom_funcs.rs:296]: These nested matches can be collapsed.

**Comment by nahla-nee** [comment]
I'll get working on those changes. It just occurred to me that I didn't add an api for `get_sys_info_ns`. Do you want that added as well?

**Comment by thejpster** [comment]
> I'll get working on those changes. It just occurred to me that I didn't add an api for `get_sys_info_ns`. Do you want that added as well?

I guess it would be good to have that, although I haven't written much NS software for the 2350.

File: rp235x-hal/src/rom_data.rs [hunk 6]
```diff
@@ -109,6 +109,530 @@ pub fn rom_data_lookup(tag: RomFnTableCode, mask: u32) -> usize {
...
+                let mut buffer = [0u32; FLAG.buffer_length()];
+                let result =
+                    unsafe { super::get_sys_info(buffer.as_mut_ptr(), buffer.len(), FLAG as u32) };
+
+                if result < 0 {
+                    return Err(BootRomApiErrorCode::from(result));
+                }
+                // The operation returned successfully but the flag wasn't supported
+                // for one reason or another
+                else if buffer[0] == 0 {
```
Comment (thejpster) [inline rp235x-hal/src/rom_data.rs:592]:
```suggestion
                else if buffer[0] == 0 {
                    // The operation returned successfully but the flag wasn't supported
                    // for one reason or another
```

File: rp235x-hal/src/rom_data.rs [hunk 7]
```diff
@@ -109,6 +109,530 @@ pub fn rom_data_lookup(tag: RomFnTableCode, mask: u32) -> usize {
...
+                const FLAG: GetSysInfoFlag = $flag;
+                const BUFFER_LEN: usize = FLAG.buffer_length();
+                let mut buffer = [0u32; FLAG.buffer_length()];
+                let result =
+                    unsafe { super::get_sys_info(buffer.as_mut_ptr(), buffer.len(), FLAG as u32) };
+
+                if result < 0 {
+                    return Err(BootRomApiErrorCode::from(result));
+                }
+                else if buffer[0] == 0 {
```
Comment (thejpster) [inline rp235x-hal/src/rom_data.rs:590]:
```suggestion
                } else if buffer[0] == 0 {
```

**Review by thejpster — APPROVED** [review]

**Comment by thejpster** [comment]
Looks great! Thank you.

**Comment by thejpster** [comment]
ah, needs a `cargo fmt` please

**Comment by nahla-nee** [comment]
I went ahead and merged the NS branch into my current branch. Hopefully that's not an issue, I glanced at the notification I got about your message which cut off the part about it not being a blocker. And cargo fmt has been ran 😅 forgot i made changes to the examples folder and only ran it in the hal folder

File: rp235x-hal/src/rom_data.rs [hunk 8]
```diff
@@ -109,6 +109,604 @@ pub fn rom_data_lookup(tag: RomFnTableCode, mask: u32) -> usize {
...
+                let mut buffer = [0u32; FLAG.buffer_length()];
+                let result =
+                    unsafe { super::get_sys_info_ns(buffer.as_mut_ptr(), buffer.len(), FLAG as u32) };
+
+                if result < 0 {
+                    return Err(BootRomApiErrorCode::from(result));
+                }
+                // The operation returned successfully but the flag wasn't supported
+                // for one reason or another
+                else if buffer[0] == 0 {
```
Comment (thejpster) [inline rp235x-hal/src/rom_data.rs:628]:
```suggestion
                } else if buffer[0] == 0 {
                    // The operation returned successfully but the flag wasn't supported
                    // for one reason or another
```

File: rp235x-hal/src/rom_data.rs [hunk 9]
```diff
@@ -109,6 +109,603 @@ pub fn rom_data_lookup(tag: RomFnTableCode, mask: u32) -> usize {
...
+    /// Generates a function with the following signature:
+    ///
+    /// ```rs
+    /// pub fn $function_name() -> Result<Option<$ok_ret_type>, BootRomApiErrorCode>
+    /// ```
+    ///
+    /// Which safely calls [`get_sys_info_ns`](super::get_sys_info_ns()) using the flag provided via
+    /// the `flag` argument. `flag` is an expression that must resolve to a const variant of
+    /// [`GetSysInfoFlag`]
+    macro_rules! declare_get_sys_info_ns_function {
```
Comment (thejpster) [inline rp235x-hal/src/rom_data.rs:613]: This macro is never used?

Comment (nahla-nee) [reply rp235x-hal/src/rom_data.rs:613]: should be fixed now

**Comment by thejpster** [comment]
Before:

```text
get_sys_info(CHIP_INFO/0x0001) -> 4
        Supported Flags: 0x0001
        RP2350 Package ID: 0x00000000
        RP2350 Device ID : 0x2d92c1e1
        RP2350 Wafer ID  : 0xb6d90d05
get_sys_info(CPU_INFO/0x0004) -> 2
        Supported Flags: 0x0004
        CPU Architecture: Arm
get_sys_info(FLASH_DEV_INFO/0x0008) -> 2
        Supported Flags: 0x0008
        CS0 Size: 16M
        CS1 Size: None
get_sys_info(BOOT_RANDOM/0x0010) -> 5
        Supported Flags: 0x0010
        A random number: 0xcfb87af017b6a85f4c1de260cef281e3
```

After:

```text
get_sys_info(CHIP_INFO/0x0001)
        RP2350 Package ID: 0x00000000
        RP2350 Device ID : 0x2d92c1e1
        RP2350 Wafer ID  : 0xb6d90d05
get_sys_info(CPU_INFO/0x0004)
        CPU Architecture: Arm
get_sys_info(FLASH_DEV_INFO/0x0008)
        CS1 GPIO: 0
        D8H Erase Supported: false
        CS0 Size: 16M
        CS1 Size: None
get_sys_info(BOOT_RANDOM/0x0010)
        A random number: 0x6713d047488e82a24dc9660817f0f669
```

Looks great!

**Pull request merged**